apiserver/secretstorage/storage.go

```diff
@@ -0,0 +1,406 @@
+	}
+
+	return &secretStorage{
+		object:    object,
+		client:    cc,
+		codec:     codec,
+		namespace: backingNS,
+	}, nil
+}
+
+// New implements rest.Storage
+func (s *secretStorage) New() runtime.Object {
+	return s.object.New()
+}
+
+func (s *secretStorage) NewList() runtime.Object {
+	return s.object.NewList()
+}
+
+// Destroy implements rest.Storage
+func (s *secretStorage) Destroy() {}
+
+// NamespaceScoped implements rest.Scoper
+func (s *secretStorage) NamespaceScoped() bool {
+	return s.object.NamespaceScoped()
+}
+
+func (s *secretStorage) Create(ctx context.Context, obj runtime.Object, createValidation rest.ValidateObjectFunc, opts *metav1.CreateOptions) (runtime.Object, error) {
+	return s.create(ctx, obj, createValidation, opts)
+}
+
+func (s *secretStorage) ConvertToTable(ctx context.Context, obj runtime.Object, tableOptions runtime.Object) (*metav1.Table, error) {
+	return rest.NewDefaultTableConvertor(s.object.GetGroupVersionResource().GroupResource()).ConvertToTable(ctx, obj, tableOptions)
+}
+
+func (s *secretStorage) Delete(ctx context.Context, name string, deleteValidation rest.ValidateObjectFunc, options *metav1.DeleteOptions) (runtime.Object, bool, error) {
+	obj, err := s.Get(ctx, name, &metav1.GetOptions{})
+	if err != nil {
+		return nil, false, err
+	}
+
+	if deleteValidation != nil {
+		if err := deleteValidation(ctx, obj); err != nil {
+			return nil, false, fmt.Errorf("failed to validate object: %w", err)
+		}
+	}
+
+	return obj, true, s.client.Delete(ctx, &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: s.getBackingNamespace(),
+		},
+	}, &client.DeleteOptions{
+		GracePeriodSeconds: options.GracePeriodSeconds,
+		Preconditions:      options.Preconditions,
+		PropagationPolicy:  options.PropagationPolicy,
+		DryRun:             options.DryRun,
+	})
+}
+
+func (s *secretStorage) DeleteCollection(ctx context.Context, deleteValidation rest.ValidateObjectFunc, options *metav1.DeleteOptions, listOptions *metainternalversion.ListOptions) (runtime.Object, error) {
+	return nil, fmt.Errorf("not implemented")
+}
+
+func (s *secretStorage) Get(ctx context.Context, name string, options *metav1.GetOptions) (runtime.Object, error) {
+	rs := corev1.Secret{}
+
+	if err := s.client.Get(ctx, client.ObjectKey{Name: name, Namespace: s.getBackingNamespace()}, &rs); err != nil {
+		// Wrapping the not found error breaks kubectl apply (404)
+		return nil, err
+	}
+
+	return s.objectFromBackingSecret(&rs)
+}
+
+func (s *secretStorage) List(ctx context.Context, options *metainternalversion.ListOptions) (runtime.Object, error) {
+	rsl := &corev1.SecretList{}
+	if err := s.client.List(ctx, rsl, &client.ListOptions{
+		// TODO(swi): Should we add labels to the secret so we can filter on them?
+		// Those would need to be hashed to not introduce collisions with controllers tracking state through labels like argocd does.
+		// LabelSelector:         nil,
+		// FieldSelector:         nil,
```

Comment:
Sounds like a good idea. I'd propose to open a ticket for it instead of a TODO, as TODOs are usually never actually done

Reply:
I submitted an issue: https://github.com/appuio/control-api/issues/100